pytket/extensions/cutensornet/mps/mps_gate.py

```diff
@@ -219,102 +216,126 @@ def _apply_2q_gate(self, positions: tuple[int, int], gate: Op) -> MPSxGate:
                 new_dim += 1
+            this_fidelity = numer / denom
 
-            # We need to change the shape of the tensors
+            # Reshape tensors down to `new_dim` for the virtual bond
+            # No data is copied or moved around, we're changing the ndarray bounds
             l_shape[-2] = new_dim
-            L.data = cp.empty(l_shape, dtype=self._complex_t)
+            # pylint: disable = unexpected-keyword-arg   # Disable pylint for next line
+            L.data = cp.ndarray(
+                l_shape,
+                dtype=self._complex_t,
+                memptr=L.data.data,
+                strides=L.data.strides,
+            )
             r_shape[0] = new_dim
-            R.data = cp.empty(r_shape, dtype=self._complex_t)
-            S_d = cp.empty(new_dim, dtype=self._real_t)
-
-            # Create new tensor descriptors
-            cutn.destroy_tensor_descriptor(L_desc)
-            cutn.destroy_tensor_descriptor(R_desc)
-            L_desc = L.get_tensor_descriptor(self._libhandle)
-            R_desc = R.get_tensor_descriptor(self._libhandle)
-
-        # Configure SVD parameters
-        svd_config_attributes = [
-            # TensorSVDPartition.US asks that cuTensorNet automatically
-            # contracts the tensor of singular values (S) into one of the
-            # two tensors (U), named L in our case.
-            (
-                cutn.TensorSVDConfigAttribute.S_PARTITION,
-                cutn.TensorSVDPartition.US,
-            ),
-        ]
+            # pylint: disable = unexpected-keyword-arg   # Disable pylint for next line
+            R.data = cp.ndarray(
+                r_shape,
+                dtype=self._complex_t,
+                memptr=R.data.data,
+                strides=R.data.strides,
+            )
+            # pylint: disable = unexpected-keyword-arg   # Disable pylint for next line
+            S_d = cp.ndarray(new_dim, dtype=self._real_t, memptr=S_d.data)
 
-        if truncating:
-            # Renormalise after truncation. Thanks to using canonical form of
-            # the MPS and the fact that the original state is a unit vector,
-            # we know that the L2 norm of the singular values (i.e. sum of its
-            # squares) must be equal to 1. We ask cuTensorNet to renormalise
-            # the truncated S so that this is satisfied after truncation, thus
-            # making sure the resulting state is normalised.
-            svd_config_attributes.append(
+            # Normalise
+            S_d *= np.sqrt(1 / this_fidelity)
+
+            # Contract S into L
+            S_d = S_d.astype(dtype=self._complex_t, copy=False)
+            v_bond = L.bonds[-2]
+            # Use some einsum index magic: since `v_bond` appears in the
+            # list of bonds of the output, it is not summed over.
+            # This causes S_d to act as the intended diagonal matrix.
+            L.data = cq.contract(L.data, L.bonds, S_d, [v_bond], L.bonds)

```

Comment:
Isn't S_d already a diagonal matrix? (I am not saying something might be wrong just trying to understand what is this "magic" doing).

Reply:
The shape of S_d is just `shape = (new_dim,)`, it's a rank-1 tensor. A "diagonal matrix" would be a rank-2 tensor with a lot of zero coefficients. If you look at the inputs to `cq.contract` here, `v_bond` appears in the list of `L.bonds` so it's both appearing in the contracted bonds **and** in the output; that's where the magic is happening.

Reply:
Oh... I see! Thank you for explaining!